### ISSUE TITLE
[4.0][Improvement Proposal] select2_with_ajax delay 500ms before doing the AJAX query

### ISSUE THREAD

**Issue by tabacitu** [Laravel-Backpack/CRUD#2387]
Check out https://stackoverflow.com/questions/59536539/laravel-backpack-select2-from-ajax-without-pagination?noredirect=1#comment105728365_59536539 for a full description.

But the gist of it is: **on big tables, select2_with_ajax takes a long time to get results**, because it makes a query after each letter typed. It's been reported that changing it to 500ms delay offers much betters results. So maybe we should change it as such? What do people think? Any downsides to it?

**Comment by danbka33** [comment]
not always used to work with large tables, the best solution is optional timeout.

**Comment by pxpm** [comment]
Hello guys. Hope you are doing well.

I read the problem in stackoverflow and I think there are two scenarios here. 

## 1

##### We want to be able to return non-paginated entries from the ajax endpoint 

- Currently `FetchOperation` specificaly the `fetch.blade.php` already supports paginated and non-paginated results.

## 2

##### For huge databases use 500ms delay between searches.

- I have implemented this in #2504 

I think to allow non-paginated results in `select2 ajax` fields we just need to change the way we add those options to the select like I did on fetch:

```php
//if we have data.data it's a paginated result, otherwise is a collection of results
//if you want to return a single result you can `return collect(Article::first())`
 if(data.data) {
                        var result = {
                            results: $.map(data.data, function (item) {
                                return {
                                    text: item[$fieldAttribute],
                                    id: item[$connectedEntityKeyName]
                                }
                            }),
                           pagination: {
                                 more: data.current_page < data.last_page
                           }
                        };
                        }else {
                            var result = {
                                results: $.map(data, function (item) {
                                    return {
                                        text: item[$fieldAttribute],
                                        id: item[$connectedEntityKeyName]
                                    }
                                }),
                                pagination: {
                                    more: false,
                                }
                            }
                        }
```

Our plan is to discontinue select2_ajax in favor of fetch, but should this be provided in 4.0 for select2 ajax fields ? 

Best,
Pedro

**Comment by tabacitu** [comment]
Huh good question @pxpm . I guess it doesn't hurt to have it inside ```select2_ajax``` too, since it's already done, and it's a non-breaking change 😄 

Let's close this issue - I'll be merging this into ```master```, so it'll be a ```4.0``` feature.
